src/functional/ffront/decorator.py

```diff
@@ -0,0 +1,294 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import collections
+import dataclasses
+import functools
+import types
+import typing
+from typing import Any, Optional, Protocol
+
+from eve.utils import UIDs
+from functional.common import GTTypeError
+from functional.ffront import common_types as ct
+from functional.ffront import field_operator_ast as foast
+from functional.ffront import program_ast as past
+from functional.ffront import symbol_makers
+from functional.ffront.foast_to_itir import FieldOperatorLowering
+from functional.ffront.func_to_foast import FieldOperatorParser
+from functional.ffront.func_to_past import ProgramParser
+from functional.ffront.past_passes.type_deduction import ProgramTypeDeduction
+from functional.ffront.past_to_itir import ProgramLowering
+from functional.ffront.source_utils import ClosureRefs
+from functional.iterator import ir as itir
+from functional.iterator.backend_executor import execute_program
+
+
+DEFAULT_BACKEND = "roundtrip"
+
+
+@typing.runtime_checkable
+class GTCallable(Protocol):
+    """Protocol (abstract base class) defining the interface for callables.
+
+    Direct subclasses of this base class only need to implement the
+    abstract methods to be callable from inside a program or from other
+    callables.
+    """
+
+    def __gt_type__(self) -> ct.FunctionType:
+        """Return symbol type, i.e. signature and return type.
+
+        The type is used internally to populate the closure vars of the
+        various dialects root nodes (i.e. FOAST Field Operator, PAST Program)
+        """
+        ...
+
+    def __gt_itir__(self) -> itir.FunctionDefinition:
```

Comment:
Either the protocol is renamed to GTOperator or the return type hint is updated, or a TODO node is added.

Reply:
A `GTCallable` doesn't need to be an operator it can also be a regular function in terms of the ITIR. Only when it appears inside a stencil closure it needs to be an operator.